libvast/src/system/pivoter.cpp

```diff
@@ -0,0 +1,159 @@
+        return i;
+      }
+    }
+  }
+#else
+  std::string edge;
+  VAST_TRACE(st.self, VAST_ARG(st.target), VAST_ARG(indicator.name()));
+  if (detail::starts_with(st.target, "suricata") || st.target == "pcap.packet")
+    edge = "community_id";
+  else if (detail::starts_with(st.target, "zeek")) {
+    if (detail::starts_with(indicator.name(), "zeek"))
+      edge = "uid";
+    else
+      edge = "community_id";
+  }
+  for (auto& i : indicator.fields) {
+    if (i.name == edge) {
+      st.cache.insert({indicator, i});
+      return i;
+    }
+  }
+#endif
+  st.cache.insert({indicator, caf::none});
+  VAST_WARNING(st.self, "got slice without shared column:", indicator.name());
+  return caf::none;
+}
+
+} // namespace
+
+pivoter_state::pivoter_state(caf::event_based_actor*) {
+  // nop
+}
+
+caf::behavior pivoter(caf::stateful_actor<pivoter_state>* self, caf::actor node,
+                      std::string target, expression expr) {
+  auto& st = self->state;
+  st.self = self;
+  st.node = node;
+  st.expr = std::move(expr);
+  st.target = std::move(target);
+  auto quit_if_done = [=]() {
+    auto& st = self->state;
+    if (st.initial_completed && st.outstanding_requests == 0)
+      self->quit();
+  };
+  self->set_down_handler([=](const caf::down_msg& msg) {
+    // Only the spawned EXPORTERs are expected to send down messages.
+    auto& st = self->state;
+    st.outstanding_requests--;
+    VAST_DEBUG(self, "received DOWN from", msg.source,
+               "outstanding requests:", st.outstanding_requests);
+    quit_if_done();
+  });
+  return {[=](vast::table_slice_ptr slice) {
+            auto& st = self->state;
+            auto pivot_field = common_field(st, slice->layout());
+            if (!pivot_field)
+              return;
+            VAST_DEBUG(self, "uses", *pivot_field, "to extract", st.target,
+                       "events");
+            auto pivot_field_name = pivot_field->name;
+            auto column = slice->column(pivot_field_name);
+            std::stringstream sstr;
+            sstr << "#type == \"" << st.target << "\" && " << pivot_field_name
+                 << " in [\"";
+            size_t request_size = 0;
+            for (size_t i = 0; i < column->rows(); ++i) {
+              auto data = materialize((*column)[i]);
+              auto x = caf::get_if<std::string>(&data);
+              // Skip if no value
+              if (!x)
+                continue;
+              // Skip if id was already requested
+              if (st.requested_ids.count(*x))
+                continue;
+              if (request_size != 0)
+                sstr << "\", \"";
+              sstr << *x;
+              st.requested_ids.insert(*x);
+              request_size++;
+            }
+            VAST_DEBUG(self, "queries for", request_size, pivot_field_name);
+            if (!request_size) {
+              return;
+            }
+            sstr << "\"]";
+            VAST_DEBUG(self, "sends query", sstr.str());
```

Comment:
I'd a bit more context about the recipient, e.g., `"sends query to node"` or `"spawns new exporter with query"`.